components/server/src/ome/security/basic/BasicACLVoter.java

```diff
@@ -84,12 +89,27 @@
 
     protected final Roles roles;
 
+    /* ignored if empty set */
+    private Set<Class<? extends IObject>> chgrpPermittedClasses = Collections.emptySet();
+
+    /* ignored if empty set */
+    private Set<Class<? extends IObject>> chownPermittedClasses = Collections.emptySet();
+
     private final LightAdminPrivileges adminPrivileges;
 
+    /* preserve permissions past context invalidation through to post-processing */
+    private final ThreadLocal<Set<AdminPrivilege>> adminPrivilegesCache = new ThreadLocal<>();
```

Comment:
Not as a must, but if possible I'd suggest concentrating `ThreadLocal` instances in as few places as possible. Here it would most likely be `CurrentDetails`:

```
$ git grep new\ ThreadLocal | grep -v OmeroCpp | grep -v TEST
components/server/src/ome/security/SecurityFilterHolder.java:    protected ThreadLocal<SecurityFilter> current = new ThreadLocal<SecurityFilter>() {
components/server/src/ome/security/SecuritySystemHolder.java:    protected ThreadLocal<SecuritySystem> current = new ThreadLocal<SecuritySystem>() {
components/server/src/ome/security/basic/CurrentDetails.java:    private final ThreadLocal<LinkedList<BasicEventContext>> contexts = new ThreadLocal<LinkedList<BasicEventContext>>();
components/server/src/ome/security/basic/CurrentDetails.java:        = new ThreadLocal<Map<String, String>>();
components/server/src/ome/tools/hibernate/ProxyCleanupFilter.java:        private final ThreadLocal<Integer> depth = new ThreadLocal<Integer>() {
```